File: frontend/src/app/features/transactions/models/transaction.ts
import { Account } from '../../../features/accounts/models/account';
import { Category } from '../../../features/categories/models/category';

export interface Transaction {
  id: number;
  title: string;
  date: string;        // YYYY-MM-DD
  credit: boolean;     // true if credit, false if debit
  amount: number;
  description: string | undefined;
  category: Category | undefined
  account: Account | undefined;
}

export interface TransactionDTO {
  title: string;
  date: string;        // YYYY-MM-DD
  credit: boolean;     // true if credit, false if debit
  amount: number;
  description: string | undefined;
  categoryId: number | undefined
  accountId: number | undefined;
}

File: backend/src/main/java/com/dallinjohnson/financeManagerAPI/model/Transaction.java
package com.dallinjohnson.financeManagerAPI.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Data;

import java.math.BigDecimal;
import java.time.LocalDate;

@Entity
@Data
public class Transaction {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "transaction_seq")
    @SequenceGenerator(name = "transaction_seq", sequenceName = "transaction_seq", allocationSize = 1)
    private Long id;

    @Column(nullable = false)
    private boolean isCredit = false;

    @Column(nullable = false)
    @NotNull(message = "Transaction date is required")
    private LocalDate date;

    @Column(precision = 15, scale = 2, nullable = false)
    @NotNull(message = "Transaction amount is required")
    @DecimalMin(value = "0.01", message = "Transaction amount must be at least 0.01")
    @DecimalMax(value = "1000000.00", message = "Transaction amount cannot exceed 1,000,000.00")
    private BigDecimal amount;

    @Size(max = 255, message = "Transaction description cannot exceed 255 characters")
    private String description;

    @ManyToOne
    @JoinColumn(name = "category_id")
    private Category category;

    @ManyToOne
    @JoinColumn(name = "account_id")
    private Account account;

    @ManyToOne
    @JoinColumn(name = "user_id")
    @JsonIgnore
    private User user;
}
